### PR TITLE
PM-396 - Db storage amazon keyspaces (cassandra)

### PR DESCRIPTION
**Pull request by piotr-iohk**
This PR:
- provides option to store blocks and submissions in AWS Keyspaces Cassandra database
- provides migration scripts to create database `make db-migrate-up` and `make db-migrate-down`
- adds integration test checking that program successfuly saves data to AWS Keyspaces Cassandra database
- documents configuration of aws keyspaces backend (also providing missing documentation about other configs AWS S3 and local filesystem)

### CODE REVIEW

File: database/cert/sf-class2-root.crt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+zt0fhvRbVazc1xDCDqmI56FspGowaDELMAkGA1UEBhMCVVMxJTAjBgNVBAoTHFN0
+YXJmaWVsZCBUZWNobm9sb2dpZXMsIEluYy4xMjAwBgNVBAsTKVN0YXJmaWVsZCBD
+bGFzcyAyIENlcnRpZmljYXRpb24gQXV0aG9yaXR5ggEAMAwGA1UdEwQFMAMBAf8w
+DQYJKoZIhvcNAQEFBQADggEBAAWdP4id0ckaVaGsafPzWdqbAYcaT1epoXkJKtv3
+L7IezMdeatiDh6GX70k1PncGQVhiv45YuApnP+yz3SFmH8lU+nLMPUxA2IGvd56D
+eruix/U0F47ZEUD0/CwqTRV/p2JdLiXTAAsgGh1o+Re49L2L7ShZ3U0WixeDyLJl
+xy16paq8U4Zt3VekyvggQQto8PT7dL5WXXp59fkdheMtlb71cZBDzI0fmgAKhynp
+VSJYACPq4xJDKVtHCN2MQWplBqjlIapBtJUhlbl90TSrE9atvNziPTnNvT51cKEY
+WQPJIrSPnNVeKtelttQKbfi3QBFGmh95DmK/D5fs4C8fF5Q=
+-----END CERTIFICATE-----
```
Comment (Sventimir) [inline database/cert/sf-class2-root.crt:24]: What is the purpose of this certificate? Why is it being kept in repo? Shouldn't it be provided by the people executing the program?

File: database/migrations/1_create_submissions_table.up.cql [hunk 2]
```diff
@@ -0,0 +1,13 @@
+CREATE TABLE IF NOT EXISTS submissions (
+    submitted_at_date TEXT,
+    submitted_at TIMESTAMP,
+    submitter TEXT,
+    created_at TIMESTAMP,
```
Comment (Sventimir) [inline database/migrations/1_create_submissions_table.up.cql:5]: The difference between `submitted_at` and `created_at` is not obvious. Also why `created_at` does not have an associated date part? Could you try renaming the columns to make this less ambiguous? If impossible, let's at the very least add a comment explaining the difference here.

File: database/migrations/2_create_blocks_table.up.cql [hunk 3]
```diff
@@ -0,0 +1,4 @@
+CREATE TABLE IF NOT EXISTS blocks (
+    block_hash TEXT PRIMARY KEY,
+    raw_block BLOB
+);
```
Comment (Sventimir) [inline database/migrations/2_create_blocks_table.up.cql]: We would probably like to keep some block metadata separately for easier access, but at the moment I don't yet know what metadata is required by the verification tool. Let's stay flexible with this table, allowing for easy addition of columns in the future, ok?

**Review by Sventimir — APPROVED** [review]
Looks okay to me. Just some minor remarks below.

File: database/migrations/2_create_blocks_table.up.cql (hunk 3 above)
Comment (piotr-iohk) [reply database/migrations/2_create_blocks_table.up.cql]: There is a database migration mechanism in place, so adding new columns should not be a problem.

File: database/migrations/1_create_submissions_table.up.cql (hunk 2 above)
Comment (piotr-iohk) [reply database/migrations/1_create_submissions_table.up.cql:5]: I wanted to be in par with all the info that is kept in the AWS S3, and this one contains both `submitted_at` (with also a `submitted_at_date`) and `created_at` (cf: https://github.com/MinaFoundation/uptime-service-backend#cloud-storage). So there are no changes to what is produced by the program, the change is only where the produced data is kept.

File: database/cert/sf-class2-root.crt (hunk 1 above)
Comment (piotr-iohk) [reply database/cert/sf-class2-root.crt:24]: This cert is required for connecting to AWS Keyspace. It is actually needed by the integration test. It is available publicly (https://docs.aws.amazon.com/keyspaces/latest/devguide/using_go_driver.html#using_go_driver.BeforeYouBegin) so I guess no harm keeping it in the repo for convenience.

File: database/migrations/1_create_submissions_table.up.cql (hunk 2 above)
Comment (Sventimir) [reply database/migrations/1_create_submissions_table.up.cql:5]: I see.

File: README.md [hunk 4]
```diff
@@ -54,9 +54,90 @@ Backend Service is a web server that exposes the following entrypoints:
...
+   - `CONFIG_GSHEET_ID` - Set this to your Google Sheet ID.
+   - `DELEGATION_WHITELIST_LIST` - Set this to your delegation whitelist list.
+   - `DELEGATION_WHITELIST_COLUMN` - Set this to your delegation whitelist column.
+
+2. **AWS S3 Configuration**:
+   - `AWS_ACCOUNT_ID` - Your AWS Account ID.
+   - `AWS_BUCKET_NAME_SUFFIX` - Suffix for the AWS S3 bucket name.
+   - `AWS_REGION` - The AWS region.
+   - `AWS_ACCESS_KEY_ID` - Your AWS Access Key ID.
+   - `AWS_SECRET_ACCESS_KEY` - Your AWS Secret Access Key.
```
Comment (smorci) [inline README.md]:
```suggestion
## Configuration

The program can be configured using either a JSON configuration file or environment variables. Below is the comprehensive guide on how to configure each option.

### Configuration Using a JSON File

1. **Set Configuration File Path**:
   Set the environment variable `CONFIG_FILE` to the path of your JSON configuration file.

2. **JSON Configuration Structure**:
   Your JSON file should adhere to the structure specified by the `AppConfig` struct in Go. Here is an example structure:

```json
{
  "network_name": "your_network_name",
  "gsheet_id": "your_google_sheet_id",
  "delegation_whitelist_list": "your_whitelist_list",
  "delegation_whitelist_column": "your_whitelist_column",
  // provide one of the below options
  "aws": {
    "account_id": "your_aws_account_id",
    "bucket_name_suffix": "your_bucket_name_suffix",
    "region": "your_aws_region",
    "access_key_id": "your_aws_access_key_id",
    "secret_access_key": "your_aws_secret_access_key"
  },
  "aws_keyspaces": {
    "keyspace": "your_aws_keyspace",
    "region": "your_aws_region",
    "access_key_id": "your_aws_access_key_id",
    "secret_access_key": "your_aws_secret_access_key",
    "ssl_certificate_path": "your_aws_ssl_certificate_path"
  },
  "filesystem": {
    "path": "your_filesystem_path"
  }
}
```

### Configuration Using Environment Variables

If the `CONFIG_FILE` environment variable is not set, the program will fall back to loading configuration from environment variables.

1. **General Configuration**:
   - `CONFIG_NETWORK_NAME` - Set this to your network name.
   - `CONFIG_GSHEET_ID` - Set this to your Google Sheet ID with the keys to whitelist.
   - `DELEGATION_WHITELIST_LIST` - Set this to your delegation whitelist sheet title where the whitelist keys are.
   - `DELEGATION_WHITELIST_COLUMN` - Set this to your delegation whitelist sheet column where the whitelist keys are.

2. **AWS S3 Configuration**:
   - `AWS_ACCOUNT_ID` - Your AWS Account ID.
   - `AWS_BUCKET_NAME_SUFFIX` - Suffix for the AWS S3 bucket name.
   - `AWS_REGION` - The AWS region.
   - `AWS_ACCESS_KEY_ID` - Your AWS Access Key ID.
   - `AWS_SECRET_ACCESS_KEY` - Your AWS Secret Access Key.
```

File: database/cert/sf-class2-root.crt (hunk 1 above)
Comment (smorci) [reply database/cert/sf-class2-root.crt:24]: I am not sure about this, let's have some further discussions so I can understand it better.

File: database/migrations/1_create_submissions_table.down.cql
Comment (smorci) [inline database/migrations/1_create_submissions_table.down.cql:1]: The name is confusing to me. Why not `1_drop_submissions_table.down.cql`? Do they have to have the same prefix?

File: src/delegation_backend/app_config.go [hunk 5]
```diff
@@ -84,15 +92,15 @@ func LoadEnv(log logging.EventLogger) AppConfig {
...
 	}
-	if config.Database != nil {
+	if config.AwsKeyspaces != nil {
 		configCount++
 	}
 	if config.LocalFileSystem != nil {
 		configCount++
 	}
 
 	if configCount != 1 {
```
Comment (smorci) [inline src/delegation_backend/app_config.go]: `configCount > 1` would be more appropiate here and `configCount < 1` for the case when no config is provided. Just friendly error messages.

**Pull request merged**